### PR TITLE
Cli Application Smoke Test

### PR DESCRIPTION
**Pull request by randomanderson**
This adds a smoke test that ensures CLI applications correctly exit when running under the java agent.  There are 2 parts:

1. A simple cli application that quits after a delay
2. The smoke test that checks to make sure that the cli application exits

I also manually tested and ensured that the test failed when JMXFetch.daemon = false.  This was the issue that prompted the creation of the test.

### CODE REVIEW

File: dd-smoke-tests/cli/src/test/groovy/datadog/smoketest/CliApplicationSmokeTest.groovy [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    ProcessBuilder processBuilder = new ProcessBuilder(command)
+    processBuilder.directory(new File(buildDirectory))
+  }
+
+  def "Cli application process ends before timeout"() {
+    setup:
+    def conditions = new PollingConditions(timeout: INSTRUMENTATION_DELAY, initialDelay: SHUTDOWN_DELAY)
+
+    expect:
+    serverProcess.isAlive()
```
Comment (mar-kolya) [inline dd-smoke-tests/cli/src/test/groovy/datadog/smoketest/CliApplicationSmokeTest.groovy]: This is 'racy'. You may want to consider making your test app exist with some 'special' exit code on success and check for it.

File: dd-smoke-tests/cli/src/main/java/datadog/smoketest/cli/CliApplication.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+/** Simple application that sleeps for a bit than quits. */
+public class CliApplication {
+
+  /** @param args Only argument is the sleep delay (in seconds) */
+  public static void main(final String[] args) throws InterruptedException {
+    final int delay = Integer.parseInt(args[0]);
+
+    System.out.println("Going to shut down after " + delay + "seconds");
+
+    Thread.sleep(delay * 1000);
```
Comment (mar-kolya) [inline dd-smoke-tests/cli/src/main/java/datadog/smoketest/cli/CliApplication.java]: You may want to have some loop with with some `@Trace` or something to make sure tracer actually gets triggered and all things are initialized.

Comment (randomanderson) [reply dd-smoke-tests/cli/src/main/java/datadog/smoketest/cli/CliApplication.java]: Testing that traces were written is not quite possible with the current setup of the smoke tests.  That said, I added some more to the CliApplication so its closer to the original application that exposed the shutdown issue: making a request using URLConnection then quitting.

File: dd-smoke-tests/cli/src/test/groovy/datadog/smoketest/CliApplicationSmokeTest.groovy (hunk 1 above)
Comment (randomanderson) [reply dd-smoke-tests/cli/src/test/groovy/datadog/smoketest/CliApplicationSmokeTest.groovy]: Agreed.  A correct exit code covers both start and stop of the process.

`isAlive()` isn't available in JDK7 anyway.

File: dd-smoke-tests/cli/src/main/java/datadog/smoketest/cli/CliApplication.java (hunk 2 above)
Comment (mar-kolya) [reply dd-smoke-tests/cli/src/main/java/datadog/smoketest/cli/CliApplication.java]: > Testing that traces were written is not quite possible with the current setup of the smoke tests.

You could potentially fire up a server and pass it as an agent url... but maybe not this time :)

**Comment by tylerbenson** [comment]
If you comment out this line, does the test fail: https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/agent-jmxfetch/src/main/java/datadog/trace/agent/jmxfetch/JMXFetch.java#L78

**Comment by randomanderson** [comment]
@tylerbenson Yes commenting out `.daemon(true)` or setting `.daemon(false)` in JMXFetch both fail the test.

@labbati 
I did add a sleep in the cliapplication because (rarely) the test didn't fail when it should. I assume it was because the JMX Processes didn't fire up completely as you mentioned.  With the sleep it always failed when it was supposed to.

File: dd-smoke-tests/cli/src/main/java/datadog/smoketest/cli/CliApplication.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    app.makeRequest();
+
+    System.out.println("Finished making request");
+  }
+
+  @Trace(operationName = "example")
+  public void makeRequest() {
+    try {
+      final URL url = new URL("http://www.example.com/");
+      final URLConnection connection = url.openConnection();
```
Comment (tylerbenson) [inline dd-smoke-tests/cli/src/main/java/datadog/smoketest/cli/CliApplication.java]: I don't like making actual requests to external URLs for testing.  How important is this?

Comment (randomanderson) [reply dd-smoke-tests/cli/src/main/java/datadog/smoketest/cli/CliApplication.java]: Not important at all.  Originally, I just had a `Thread.sleep()`.  @mar-kolya commented that the application wasn't doing enough so I added the request.  The test doesn't rely on the success or failure of the request.

The example gist that first showed the JMXFetch issue had a request to google.com but I don't think was necessary to show the problem.

I'm fine removing the request and going back to only having a `sleep`.

Comment (tylerbenson) [reply dd-smoke-tests/cli/src/main/java/datadog/smoketest/cli/CliApplication.java]: I think I would prefer that.  Sorry for the back-and-forth.  @mar-kolya you ok with that?

**Pull request merged**